csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -114,98 +114,561 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
+        {
+            return Run(inputNames, inputValues, outputNames, _builtInRunOptions);
+        }
+
+        /// <summary>
+        /// Runs the loaded model for the given inputs, and fetches the specified outputs in <paramref name="outputNames"/>. Uses the given RunOptions for this run.
+        /// </summary>
+        /// <param name="inputNames">Specify a collection of string that indicates the input names. Should match <paramref name="inputValues"/>.</param>
+        /// <param name="inputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the input values.</param>
+        /// <param name="outputNames">Specify a collection of string that indicates the output names to fetch.</param>
+        /// <param name="options"></param>
+        /// <returns>Output Tensors in a Collection of NamedOnnxValue. User must dispose the output.</returns>
+        public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(
+            IReadOnlyCollection<string> inputNames,
+            IReadOnlyCollection<FixedBufferOnnxValue> inputValues,
+            IReadOnlyCollection<string> outputNames,
+            RunOptions options)
+        {
+            if (inputNames.Count != inputValues.Count)
+            {
+                throw new ArgumentException($"Length of {nameof(inputNames)} ({inputNames.Count}) must match that of {nameof(inputValues)} ({inputValues.Count}).");
+            }
 
-                    inputIndex++;
+            // prepare inputs
+            string[] inputNamesArray = inputNames as string[] ?? inputNames.ToArray();
+            IntPtr[] inputValuesArray = new IntPtr[inputNames.Count];
+            int inputIndex = 0;
+            foreach (var input in inputValues)
+            {
+                inputValuesArray[inputIndex] = input.Value;
+
+                inputIndex++;
+            }
+
+            // prepare outputs
+            string[] outputNamesArray = outputNames as string[] ?? outputNames.ToArray();
+            IntPtr[] outputValuesArray = new IntPtr[outputNames.Count];
+
+            IntPtr status = NativeMethods.OrtRun(
+                                                _nativeHandle,
+                                                options.Handle,
+                                                inputNamesArray,
+                                                inputValuesArray,
+                                                (UIntPtr)inputNames.Count,
+                                                outputNamesArray,
+                                                (UIntPtr)outputNames.Count,
+                                                outputValuesArray /* Empty array is passed in to receive output OrtValue pointers */
+                                                );
+
+            try
+            {
+                NativeApiStatus.VerifySuccess(status);
+                var result = new DisposableList<DisposableNamedOnnxValue>();
```

Comment:
Can we set the capacity via the ctor for the list ? It seems like we know the count of the elements that are going to be added and hence it seems optimal before insertion in the following step...
(This propbably applies to the other Run() overrides that return a list of disposed values)